Add pickpockets to loot tracker

danielcimento opened the pull request:
Suggested a few times in the past (e.g. https://github.com/runelite/runelite/issues/5033).

I thought this would be a good addition to the loot tracker. It was something I found myself wanting when thieving master farmers. It seems to have been a bit polarizing in the past, so I've put it behind a config flag by default.

Possibly extensible to cover loot from all types of thieving (stalls, chests, etc). Let me know if there are any improvements needed

deathbeam commented:
I dont think just putting it behind config flag solves anything, the original point still stands, loot tracker is tracking monster/boss loot and thats it.

dekvall commented:
To be fair, nowadays the loot tracker also tracks Kingdom rewards, Fishing Trawler and a variety of chests. It's not exclusively tracking monster and boss loot anymore. I don't think that we need to track pickpockets of npcs which only yield coins but it would be interesting to see the loot you get from master farmers, ham members and elves etc. I think that we have passed the old line already and I am not sure where we should draw the new one. Maybe this config is a nice compromise while we decide.

abextm commented:
I think if we want a config (I'm not sure we do) it should be very limited — just [Everything/PvM]. I don't want the config for it filled with options for every single non PvM activity it can do.

danielcimento commented:
@abextm

I'd be willing to help merge everything into a non-PvM option, if you need me to.

@deathbeam

I see where you're coming from, but I'm afraid I don't quite understand why only tracking PvM loot makes the loot tracker better as a feature.

XrioBtw commented on runelite-client/src/main/java/net/runelite/client/plugins/loottracker/LootTrackerPlugin.java:
```diff
@@ -486,6 +489,15 @@ public void onChatMessage(ChatMessage event)
+		if (config.trackPickpockets() && pickpocketMatcher.matches())
+		{
+			String pickpocketTarget = WordUtils.capitalize(pickpocketMatcher.group("target"));
+			eventType = pickpocketTarget + " (Pickpocket)";
```
I would recommend adding e.g. `private static final String PICKPOCKET_EVENT = " (Pickpocket)"` and using this here and in line 542.

danielcimento replied:
Mmm, that's a good idea. I had a const at first but I took it out because I hadn't thought to use it in both places

danielcimento commented:
I've refactored everything into non-combat vs combat loot entries and changed the toggle so it hides all non-combat loot. I tried to make the change as non-intrusive as possible, though I did move a lot of the event strings/constants ("Hespori", "Herbiboar", etc) into an enum class, just to reduce clutter a bit and maybe help with future extensibility.

Let me know if these changes are amenable, or if there's something else that needs to be changed.

danielcimento commented:
Hmm, I'll take a look at those build failures

Adam- commented:
I think this is fine to just track pickpocketing. Maybe we could extend the "ignored items" feature to also ignore events if we wanted. If you want to refactor the EventType stuff it should preferably be in a separate commit prior to adding any of the new pickpocketing stuff.

danielcimento commented:
> 
> 
> I think this is fine to just track pickpocketing. Maybe we could extend the "ignored items" feature to also ignore events if we wanted. If you want to refactor the EventType stuff it should preferably be in a separate commit prior to adding any of the new pickpocketing stuff.

That makes sense. I'll move the event type stuff to a separate PR

Adam- commented:
It can be in the same PR, just separate commits. Basically just rebase and reorder them since you sort of have done it already but in reverse.

Adam- commented on runelite-client/src/main/java/net/runelite/client/plugins/loottracker/LootTrackerPlugin.java:
```diff
@@ -474,62 +467,68 @@ public void onChatMessage(ChatMessage event)
+		// H.A.M. members show up differently in the chat box, so we need to specially track them
+		if (event.getMenuOption().equals("Pickpocket"))
+		{
+			pickpocketingHamMembers = Text.removeTags(event.getMenuTarget()).equals("H.A.M. Member");
```
You could probably just store the full name here and always use it to avoid this issue in the future with other things.

danielcimento replied:
Store the name of the last pickpocket target? That was my first implementation, but I think it may have some issues on its own:

* there's a very small (though not non-existent) risk of race conditions, since players can select different menu options while mid-pickpocket, so the loot might potentially be recorded under the wrong name.
* some pickpocket-able NPCs have randomized names, though their pickpocketing message is the same. e.g. I believe all the Prifddinas elves have different names, but still display as "you pick the elf's pocket", so there'd be a lot of book-keeping required to map all those names.

I do agree that a specific boolean just for ham members is a bit hacky, so I'll consider more extensible alternatives before pushing the newest changes

Adam- commented on runelite-client/src/main/java/net/runelite/client/plugins/loottracker/LootEvent.java:
```diff
@@ -0,0 +1,64 @@
+package net.runelite.client.plugins.loottracker;
```
The license header should be added to all new files.

Adam- commented on runelite-client/src/main/java/net/runelite/client/plugins/loottracker/LootEvent.java:
```diff
@@ -0,0 +1,64 @@
+import java.util.Map;
+import java.util.stream.Collectors;
+
+public enum LootEvent
```
```suggestion
enum LootEvent
```

Adam- commented on runelite-client/src/main/java/net/runelite/client/plugins/loottracker/LootEvent.java:
```diff
@@ -0,0 +1,64 @@
+	LARRANS_CHEST_SMALL("Larran's small chest"),
+	THEATRE_OF_BLOOD("Theatre of Blood");
+
+	final String asText;
```
```suggestion
	@Getter
	private final String name;
```

Adam- commented on runelite-client/src/main/java/net/runelite/client/plugins/loottracker/LootEvent.java:
```diff
@@ -0,0 +1,64 @@
+		.put("master", CLUE_SCROLL_MASTER)
+		.build();
+
+	LootEvent(String asText)
```
Use `@RequiredArgsConstructor`

Adam- commented on runelite-client/src/main/java/net/runelite/client/plugins/loottracker/LootTrackerPlugin.java:
```diff
@@ -125,6 +130,17 @@
 	// Last man standing map regions
 	private static final Set<Integer> LAST_MAN_STANDING_REGIONS = ImmutableSet.of(13658, 13659, 13914, 13915, 13916);
 
+	private static final Pattern pickpocketingRegex = Pattern.compile("You pick (the )?(?<target>.+)'s? pocket.*");
```
```suggestion
	private static final Pattern PICKPOCKET_REGEX = Pattern.compile("You pick (the )?(?<target>.+)'s? pocket.*");
```

Adam- commented on runelite-client/src/main/java/net/runelite/client/plugins/loottracker/LootTrackerPlugin.java:
```diff
@@ -125,6 +130,17 @@
+	 * Note that if the two NPCs can be found in the same place, there is a chance of race conditions
+	 * occurring when changing targets mid-pickpocket, in which case a different solution may need to be considered.
+	 */
+	private static final Map<String, List<String>> PICKPOCKET_DISAMBIGUATION_MAP = ImmutableMap.of(
```
This can probably just be a multimap

dabolink commented:
> I think this is fine to just track pickpocketing. Maybe we could extend the "ignored items" feature to also ignore events if we wanted. If you want to refactor the EventType stuff it should preferably be in a separate commit prior to adding any of the new pickpocketing stuff.

I did create a PR a while ago #9752 which allows the user to hide specific events. Haven't checked on it but if you want I could rebase it.

Adam- commented:
I applied the changes I left a few days ago, and started a test. I noticed that Paladins aren't on this. Is that because the loot from them isn't random? I don't think it makes much sense to only track random loot.

danielcimento commented:
Yeah I originally left off Paladins because the loot wasn't random. It might be better to leave off enemies whose loot is just coin pouches instead (i.e. add paladins). It's mostly a question of whether people want it tracked. It's not too much of a hassle to add in after the fact, anyhow

Adam- commented:
can you add anything which is pickpocket-able?

danielcimento commented:
Sure

Adam- commented:
Most of this was added in #10759. I kept the `LootEvent` out as it makes the diff a lot smaller. If you want to revisit that we could, though I am not sure I want to use it for pickpocketing since it requires having to hardcode in all of the possible pickpocket targets which is not so great. The other usages seemed good, though.

Nightfirecat commented:
I am closing this PR for now, as the main objective (tracking pickpocket loots) has been implemented and merged separately. This PR, or a new PR, can be opened for the refactoring portion of this feature.
